### PR TITLE
Fix CUDA double backwards

### LINKED ISSUE #4457: Check at compile-time that grads[i] indexing in tools/autograd/derivatives.yaml are in bounds
We don't check that the grads[i] formulas in tools/autograd/derivatives.yaml match with the actual number of outputs. We already do semantic analysis on derivative expressions (to determine what to save).
I will extend this to extract the set of direct grads[i] accesses that are made and then I will check if they are in bounds, based on the metadata which specifies how many outputs the forward has.

### PR DESCRIPTION
**Pull request by ezyang**
CUDA double backwards was broken, and we didn't know about it. Fix it.

Each commit is a logical unit of work.

```
commit 9ae56c3bb4a26e01269031d4205263952ceadc79
Author: Edward Z. Yang <ezyang@fb.com>
Date:   Wed Jan 3 09:27:24 2018 -0800

    Actually test CUDA double-backwards codepath.
    
    Previously, we only tested CPU double-backwards, which is bad!
    This would have caught #4422 (still not fixed, so those tests
    are manually disabled.)
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>

commit e80a9949c19ba979b63e8faeb984b34c6929d4c2
Author: Edward Z. Yang <ezyang@fb.com>
Date:   Wed Jan 3 08:14:51 2018 -0800

    Check for out of bounds grads access in derivatives.yaml
    
    This test would have caught the OOB in thnn_conv_depthwise2d_backward
    
    Fixes #4457
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>

commit c8b34c46b508c3636d60a9bd297e755baeff7b20
Author: Edward Z. Yang <ezyang@fb.com>
Date:   Wed Jan 3 08:14:27 2018 -0800

    s/uses_grad/uses_single_grad/ for more clarity.
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>

commit 563936c3b81b4c0a37e4348d36af2265f98feabf
Author: Edward Z. Yang <ezyang@fb.com>
Date:   Wed Jan 3 09:18:18 2018 -0800

    Fix 'invalid argument 4: weight tensor has to be contiguous'
    
    Weight can be non-contiguous due to double backwards, where
    we transpose the weight.  I'm not very happy with this fix
    but it seems to make the tests pass.
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>

commit af1283c24c9d39465342654cd4b9b2eca6d1d9c6
Author: Edward Z. Yang <ezyang@fb.com>
Date:   Wed Jan 3 08:39:04 2018 -0800

    Fix two bugs in thnn_conv_depthwise2d_backward gradient.
    
    - Out of bounds grads[2] access (thnn_conv_depthwise2d_backward
      doesn't compute bias gradient)
    
    - Groups was not set appropriately for depthwise convolution
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>
```

### CODE REVIEW

File: test/common_nn.py [hunk 1]
```diff
@@ -854,6 +856,22 @@ def test_cuda(self, test_case):
                 for cpu_d_p, gpu_d_p in zip(cpu_param[1], gpu_param[1]):
                     test_case.assertEqual(cpu_d_p, gpu_d_p, 2e-4)
 
+            # Run double-backwards on CPU and GPU and compare results
+            # TODO: BatchNorm double-backwards is currently broken, see
+            # #4422; remove this when it is fixed.
+            if self.check_gradgrad and not self.get_name().startswith('test_BatchNorm'):
+                for i in range(5):
```
Comment (colesbury) [inline test/common_nn.py]: Why is this running 5 times?

File: test/common_nn.py [hunk 2]
```diff
@@ -854,6 +856,22 @@ def test_cuda(self, test_case):
...
+                    cpu_output_t = cpu_output.data if isinstance(cpu_output, Variable) else cpu_output
+                    cpu_gradOutput = Variable(cpu_output_t.clone().bernoulli_(), requires_grad=True)
```
Comment (colesbury) [inline test/common_nn.py]: Why bernoulli?

File: test/common_nn.py [hunk 3]
```diff
@@ -854,6 +856,22 @@ def test_cuda(self, test_case):
...
+                    gpu_gradOutput = Variable(cpu_gradOutput.type('torch.cuda.FloatTensor').data, requires_grad=True)
+                    cpu_gradInput = test_case._backward(cpu_module, cpu_input, cpu_output, cpu_gradOutput, create_graph=True)
+                    gpu_gradInput = test_case._backward(gpu_module, gpu_input, gpu_output, gpu_gradOutput, create_graph=True)
+                    cpu_input.grad.sum().backward(retain_graph=True)
```
Comment (colesbury) [inline test/common_nn.py]: I think this should use `torch.autograd.grad` and not accumulate back into the parameter gradients.

File: tools/autograd/load_derivatives.py [hunk 4]
```diff
@@ -48,6 +48,14 @@ def transform_return(r):
     formula, saved_inputs = saved_variables(formula, arguments)
     formula, saved_outputs = saved_variables(formula, returns)
 
+    # Check that the referenced gradients in the formula are in bounds
+    for i in used_gradient_indices(formula):
+        if i >= len(declaration['returns']):
+            raise RuntimeError(
```
Comment (colesbury) [inline tools/autograd/load_derivatives.py]: Nice!

File: test/common_nn.py [hunk 5]
```diff
@@ -854,6 +856,22 @@ def test_cuda(self, test_case):
                 for cpu_d_p, gpu_d_p in zip(cpu_param[1], gpu_param[1]):
                     test_case.assertEqual(cpu_d_p, gpu_d_p, 2e-4)
 
+            # Run double-backwards on CPU and GPU and compare results
+            # TODO: BatchNorm double-backwards is currently broken, see
+            # #4422; remove this when it is fixed.
+            if self.check_gradgrad and not self.get_name().startswith('test_BatchNorm'):
```
Comment (gchanan) [inline test/common_nn.py]: can we explicitly write check_gradgrad=False in all the test specifications where we need it?  This kind of code leads to people realizing months later that we never turned on the tests.

File: test/common_nn.py (hunk 1 above)
Comment (ezyang) [reply test/common_nn.py]: I copied the pattern from the code above, which does the test for backward. It runs it 5 times because it generates random data and then sees if the CPU and GPU implementations line up; it's a probabilistic process.

File: test/common_nn.py (hunk 2 above)
Comment (ezyang) [reply test/common_nn.py]: Ditto as above.

File: test/common_nn.py (hunk 5 above)
Comment (ezyang) [reply test/common_nn.py]: So... would you like me to do this "correctly" or "easily"? It's not correct to put `check_gradgrad=False` in the batchnorm tests because that also disables the CPU double backward tests and those are passing on HEAD. So the correct thing is to add another variable to control this but then I decided it was too much work and added a special case (which I promise to remove...)

File: test/common_nn.py (hunk 1 above)
Comment (colesbury) [reply test/common_nn.py]: The above code does not seem like a good implementation. I don't think we get enough extra signal from running 5x to justify it. `test_nn.py` is already too slow.

File: test/common_nn.py (hunk 2 above)
Comment (colesbury) [reply test/common_nn.py]: Bernoulli doesn't seem like a good choice of random function: 0s ignore inputs and 1s can mask scaling bugs. Something like `randn` might be better.

File: test/common_nn.py (hunk 5 above)
Comment (gchanan) [reply test/common_nn.py]: Sure, add a different flag :).

File: test/common_nn.py (hunk 2 above)
Comment (ezyang) [reply test/common_nn.py]: I assume this reasoning applies to the backwards calculation too. I thought there was some clever reason bernoulli was picked but maybe not...

Comment (gchanan) [reply test/common_nn.py]: is there a reason we are (mostly) duplicating this sort of code?

Comment (ezyang) [reply test/common_nn.py]: Mostly because I was too lazy to think of a good abstraction in this case. I suppose I can write a general purpose harness for comparing two functions/modules for equivalent behavior, and another for GPU'ifying a module/input pair... if you really want me to...

File: aten/src/ATen/native/Convolution.cpp [hunk 6]
```diff
@@ -395,7 +395,10 @@ at::Tensor _convolution_nogroup(
     if (dim == 4) {
       if (dilated) {
         return at::thnn_conv_dilated2d(
-            input, weight, kernel_size, bias,
+            // TODO: This might not be necessary in the non-CUDA case,
+            // as we only triggered a failure here with
+            // test_ConvTranspose1d_cuda.  Very unsatisfactory...
+            input, weight.contiguous(), kernel_size, bias,
```
Comment (apaszke) [inline aten/src/ATen/native/Convolution.cpp]: This should be fixed in TH(CU)NN itself (add `newContiguous` for `weight`)

File: test/common_nn.py (hunk 5 above)
Comment (apaszke) [reply test/common_nn.py]: `FIXME_no_gradgrad=True`

File: test/common_nn.py (hunk 2 above)
Comment (apaszke) [reply test/common_nn.py]: That should actually be pretty easy - do the `gradgradcheck` trick. Write a function that takes another function and compares outputs and grads, then pass `torch.autograd.grad(module(x), x)` as the callable to test double backward.

Comment (ezyang) [reply test/common_nn.py]: The gradgradcheck trick is a bit difficult to do because it assumes that we are interacting with the module under test functionally, and the existing tests are not doing this.

My new plan is to delete the backward comparison code, and just run gradcheck/gradgradcheck on the CUDA kernels.

Comment (ezyang) [reply test/common_nn.py]: In offline conversation @apaszke and I came to the conclusion that it is more trouble than it is worth trying to deduplicate this code, because we have to use the stateful module API (because this code is exercised by legacy NN which doesn't support functional gradient computation), which means you can't "just" do the gradgradcheck trick, you need a version of this trick which wraps a module not a function.

Comment (ezyang) [reply test/common_nn.py]: Update: we have a different strategy, it'll be in the new patch

File: test/common_nn.py (hunk 3 above)
Comment (ezyang) [reply test/common_nn.py]: For the record, making this change helped uncover some other bugs with double backwards (#4500). So, :+1:

File: test/common_nn.py (hunk 5 above)
Comment (ezyang) [reply test/common_nn.py]: Done.

File: aten/src/ATen/native/Convolution.cpp (hunk 6 above)
Comment (ezyang) [reply aten/src/ATen/native/Convolution.cpp]: So, I took a look at THCUNN and actually, THCUNN is already running `newContiguous` for weight, and the check was added by you in 45596d52897fb187701943cb77456ff1e7249989

I am going to delete this test :)
